corehq/apps/users/audit/change_messages.py

```diff
@@ -1,91 +1,329 @@
-class UserChangeMessage(object):
+from enum import Enum
+from django.utils.translation import ugettext as _
+
+
+class Change(Enum):
+    SET = 'set'
+    RESET = 'reset'
+    REMOVE = 'remove'
+    ADD = 'add'
+
+
+class UserChangeMessageV1(object):
+    """
+    Each change message to follow the structure
+    {
+        "field": {
+            "change": change_details
+        }
+    }
+    field: could be domain, phone_numbers etc
+    change: either 'set' / 'reset' / 'remove' / 'add'
+    change_details: a Boolean or
+                    a dict with details as needed for the change
+                    a list of a dicts as needed for the change
+    """
     @staticmethod
     def program_change(program):
         if program:
-            message = f"Program: {program.name}[{program.get_id}]"
+            change_message = {"program": {Change.SET: {"id": program.get_id, "name": program.name}}}
         else:
-            message = 'Program: None'
-        return message
+            change_message = {"program": {Change.SET: {}}}
+        return change_message
 
     @staticmethod
     def role_change(user_role):
-        message = 'Role: None'
         if user_role:
-            message = f"Role: {user_role.name}[{user_role.get_qualified_id()}]"
-        return message
+            change_message = {"role": {Change.SET: {"id": user_role.get_qualified_id(), "name": user_role.name}}}
+        else:
+            change_message = {"role": {Change.SET: {}}}
```

Comment:
It could be, yes.

Here is how I have used them,
SET action would mean setting a certain value, an empty object in this case saying that no role was set.
RESET action would simply mean a reset action and would always be a boolean value (True).

Clearing role can be considered same as a "Reset" yes, though just using SET helps keeps the distinction shared above.
If there was some way we could keep just SET or RESET, that would be useful.

Reset is used for password and devices.
When password is reset it's set to a new value but we can't store that, and for devices, devices are just deleted. So "reset" sounds like the right action for them.

Does that sound reasonable?